website/client-old/js/services/undoService.js

```diff
@@ -0,0 +1,50 @@
+'use strict';
+
+angular.module('habitrpg')
+.factory('Undo', ['$rootScope', '$http', 'Tasks', 'Common',
+  function undo($rootScope, $http, Tasks, Common) {
+    var undoApi = {};
+
+    undoApi.history = [];
+
+    undoApi.addAction = function (action, data) {
+      undoApi.history.push({action: action, data: data});
+    };
+
+    undoApi.undoAction = function () {
+      var pop = undoApi.history.pop();
+      if (!pop) return;
+
+      // @TODO: We need to add a pop interface for each action
+      if (pop.action === 'scoreTask') {
+        var direction = pop.data.direction;
+
+        if (direction === "up")  {

```

Comment:
Another option: Each time a task is scored, record the change in value of the task and the things gained from it (stats, quest progress, drops). The record would be in an object in the user's document, not in the task's document. The record would include the task's ID. When the undo button is clicked, use that data to reverse the effects on the user's account and on the task (subtract the change in task value and also remove the most recent history entry when appropriate).

It could also be an array of objects if we wanted multi-undo. The array could have a "type" field so you could record skills, purchases, etc, not just task clicks. But really, I think a single undo would be enough.

Erase the object on cron so that we don't try to revert quest progress after it's already been applied to the quest.